File: crates/mithril-oxide-sys/cpp/IR/Value.cpp
#include "Value.hpp"

#include <memory>
#include <string>

#include <llvm/Support/raw_ostream.h>
#include <mlir/IR/Value.h>
#include <rust/cxx.h>


namespace mithril_oxide_sys {

rust::String Value_print(void *value)
{
    auto val = Value::getFromOpaquePointer(value);
    std::string s;
    llvm::raw_string_ostream ss(s);
    val.print(ss);
    return rust::String::lossy(s);
}

void Value_dump(void *value)
{
    auto val = Value::getFromOpaquePointer(value);
    val.dump();
}

} // namespace mithril_oxide_sys

File: crates/mithril-oxide-sys/src/IR/Value.rs
pub use self::ffi::{Value_dump, Value_print};

#[cxx::bridge]
pub(crate) mod ffi {
    #[namespace = "mithril_oxide_sys"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/IR/Value.hpp");

        type c_void;

        #[must_use]
        pub unsafe fn Value_print(value: *mut c_void) -> String;

        pub unsafe fn Value_dump(value: *mut c_void);
    }
}

#[cfg(test)]
mod tests {}
